### ISSUE TITLE
Compiled colorscheme doesn't adapt to plugin changes?

### ISSUE THREAD

**Issue by tmillr** [projekt0n/github-nvim-theme#262]
I'm not sure exactly what's going on (or the cause of this), but here's what happened. BTW I'm using `Plug`.

In my vimrc, I changed `Plug 'projekt0n/github-nvim-theme'` to `Plug '~/code/github-nvim-theme'` (i.e. I changed it to use my local fork that I'm currently developing on so that it'd be easier for me to observe and test changes that I make to the plugin's source code). I then restarted Neovim and realized that many, if not all, of the changes I had made in my fork weren't present or taking effect. To fix the issue, I had to manually run `rm ~/.cache/nvim/github-theme/github*` and then restart Neovim again.

Also, FWIW, Neovim came out with a new loader in `v0.9.0` which shims `loadfile()` to compile and cache any loaded Lua file (including, presumably, any required Lua file since `loadfile()` probably underpins `require()`). However, it is documented as experimental atm and is only available from `v0.9.0` onwards I believe. Perhaps it is worth looking-into instead? (it would cut-down on repo/plugin size and maintenance overhead if this new feature covers our use-case(s)).

**Comment by nullchilly** [comment]
> To fix the issue, I had to manually run `rm ~/.cache/nvim/github-theme/github*` and then restart Neovim again.

We only recompile if user's config changed or there was some changes in `github-nvim-theme/.git` folder. However, we can add something similar to [catppuccin/init.lua#L179-L186](https://github.com/catppuccin/nvim/blob/1d3eda15703ba70f57e94e6451db55914ff7017f/lua/catppuccin/init.lua#L179-L186) to be able to do `:GithubThemeCompile` without having to restart neovim:

```lua
vim.api.nvim_create_user_command("GithubThemeCompile", function()
	for name, _ in pairs(package.loaded) do
		if name:match "^github-theme." then package.loaded[name] = nil end
	end
	require("github-theme").compile()
	require("github-theme").load()
end, {})
```

We can add an autocmd that recompile on save if the current buffer's path matches `github-theme` if `vim.g.github_theme_debug` was on

```lua
if vim.g.github_theme_debug then
	vim.api.nvim_create_autocmd("BufWritePost", {
		pattern = "*",
		callback = function(opts)
			if opts.file:find("github-theme") then
				vim.schedule(function()
					require("github-theme").compile()
					require("github-theme").load()
				end)
			end
		end,
	})
end
```

> Neovim introduced a new Lua loader in v0.9.0.

This is what compile feature initially behaves in [nightfox.nvim](https://github.com/EdenEast/nightfox.nvim/), we relied on [impatient.nvim](https://github.com/lewis6991/impatient.nvim) (which is now known as `:h lua-loader`. However, it meant that the theme was compiled twice (One from the colorscheme and one from impatient) and increased the compile binary size so we decided to bundle it directly into the theme.

> It would cut-down on repo/plugin size, number of tests, and maintenance overhead if this new feature covers the our use-case(s).

I don't think so, the implementation is like 6 lines longer?

```
local lines = {
  [[return string.dump(function()]],
  file_content,
  [[end]]
}
local f = loadstring(table.concat(lines, "\n")
```

**Comment by tmillr** [comment]
> We only recompile if user's config changed or there was some changes in `github-nvim-theme/.git` folder. However, we can add something similar to [catppuccin/init.lua#L179-L186](https://github.com/catppuccin/nvim/blob/c9cc5a997f1dae3f35b4bdd62f35958fee363ab4/lua/catppuccin/init.lua#L179-L186) to be able to do `:GithubThemeCompile` without having to restart neovim

I see that. I don't think this will catch uncommitted changes however (nor work with `git worktree`, i.e. where `.git` is a regular file), but I guess for the majority of users who aren't going to hack on it themselves, this won't be an issue.

I believe there is already such a command provided by this repo/plugin. The thing is, I shouldn't have to manually run this command every time the plugin gets updated just for updates to take effect. Right now, at least in this repo/plugin, it seems that it doesn't work properly when I make a brand-new repo clone (which is indeed a changed git dir, no?) and completely restart Neovim. I'll investigate/test further when I get the chance, but I believe I found a bug in the implementation, and I can propose a fix soon.

> We can also add an autocmd similar to [catppuccin/nvim#L188-195](https://github.com/catppuccin/nvim/blob/c9cc5a997f1dae3f35b4bdd62f35958fee363ab4/lua/catppuccin/init.lua#L188-L195) that recompile on save if the current buffer's path matches `github-theme` if `vim.g.github_theme_debug` was on
> 
> ```lua
> if vim.g.github_theme_debug then
>   vim.api.nvim_create_autocmd("BufWritePost", {
>     pattern = "*/github-theme/*",
>     callback = function(opts)
>       vim.schedule(function()
>         require("github-theme").compile()
>         require("github-theme").load()
>       end)
>     end
>   })
> end
> ```

Thanks. I understand. I believe this repo already has something similar to this as well.

> 
>  > Neovim introduced a new Lua loader in v0.9.0.
> 
> This is what compile feature initially behaved in [nightfox.nvim](https://github.com/EdenEast/nightfox.nvim/), we relied on [impatient.nvim](https://github.com/lewis6991/impatient.nvim) (which is now known as `:h lua-loader`). However, it meant that the theme was compiled twice (One from the colorscheme and one from impatient) and increased the compile binary size so I decided to bundle it directly into [catppuccin/nvim](https://github.com/catppuccin/nvim) which later got [added back into nightfox](https://github.com/EdenEast/nightfox.nvim/pull/241).
>

I don't think bundling it is going to avoid double compilation because you have no way to know, or control, whether the user is using impatient.nvim, or the new Lua-loader, or some other solution, anyway. If I'm not mistaken, you may even be compiling unnecessarily due to this fact, which might even worsen performance (and lead to additional unnecessary files on the fs).

I believe that, in the not-so-distant future (if not now already), the proper way to handle this will be Neovim's builtin compilation, and for several reasons. Firstly, the user gets to choose what they'd like to do and only has to choose **one time** in **one central location** (instead of trying to keep several different plugin configs in-sync with each other). Secondly, as is also the case when utilizing any other feature that's already built-in to the host environment, there'll be less code (duplication), simpler/cleaner codebases, fewer tests to write and maintain, and less maintenance overhead. I mean, just look at all the headaches that packer causes due to its custom compilation feature...I see notices/warnings on every other nvim plugin's README warning about it causing problems with their plugin, and now, it seems to be causing similar issues with this plugin too. I'm assuming that Neovim's loader will fix these issues, but even if it doesn't, the user can easily disable it globally via one simple option in their vimrc.

Furthermore, is it really that much more beneficial to have just a few or so plugins capable of this feature (compiling/caching byte-code) when the majority of plugins do not? Most users have many/multiple plugins, and most of them do not have custom compilation built-in. In practice, the performance benefits are probably going to be relatively small compared to the alternative.

> > It would cut-down on repo/plugin size, number of tests, and maintenance overhead if this new feature covers the our use-case(s).
> 
> I don't think so, the implementation is like 6 lines longer?

In this repo, the current implementation for this appears to be broken. That's why I created this issue, and hence the necessity for thorough/proper testing. I suggest considering Neovim's builtin compilation instead for this reason (among others).

FWIW, the [compiler file/module](https://github.com/projekt0n/github-nvim-theme/blob/f09a14e9e111f7bfa190dbeab1994b8240dbf2ab/lua/github-theme/lib/compiler.lua) in this repo is actually 106 lines, not 6, and that's not including the additional code that it takes to use it. But it's not just about the size of the extra code/file, it's also the additional: maintenance overhead (e.g. refactoring, or writing/maintaining tests to make sure everything is working properly, both on its own, and when integrated), complexity of the codebase, number of tests, etc. Re-implementing several things that the host environment already provides leads to unnecessary maintenance overhead and duplicated code (including "duplicated" tests as these things are probably already tested upstream). It's just not as simple or clean, and it can complicate the codebase unnecessarily. That's fine if it's done for performance reasons, but only if the performance benefits are substantial enough (and worth the additional overhead).

It's not just about the compiler feature either. I came across a few things that seemed like they were re-implementing functionality that is already provided by Neovim or LuaJIT (e.g. bit operations, tables to string conversion, hashing, etc.), although maybe there is a reason for all of it that I am not aware of.

**Comment by nullchilly** [comment]
> I believe this repo already has something similar to this as well.

No, this repo is based on nightfox which I haven't ported these features to yet.

**Comment by nullchilly** [comment]
> I believe there is already such a command provided by this repo/plugin.

> I believe this repo already has something similar to this as well.

No, this repo is based on nightfox which I haven't ported these features to yet.

> If I'm not mistaken, you may even be compiling unnecessarily due to this fact, which might even worsen performance

The compiled files don't have the `*.lua` patterns so I don't think vim loader will interfere (?)

> I suggest considering Neovim's builtin compilation instead for this reason (among others).

These complications goes beyond just compiling to lua bytecode. It skips loading a bunch of modules like `modules/` `utils/` etc

> That's fine if it's done for performance reasons, but only if the performance benefits are substantial enough (and worth the additional overhead).

In the past, [catppuccin/nvim#94](https://github.com/catppuccin/nvim/issues/94) was considered a slow colorscheme because of too many features. It was slower than even nvim-cmp and people didn't like that.

Which is why we added the compiler to make startuptime goes straight from 10ms to 0.5ms

> re-implementing functionality that is already provided by Neovim or LuaJIT (e.g. bit operations, tables to string conversion, hashing, etc.)

Where is the hashing function in luajit?

Also it is an unodered table hashing which (to my knowledge) was never done before in a lua module. It was a result of hours of experimenting on my part.

**Comment by tmillr** [comment]
> > I believe there is already such a command provided by this repo/plugin.
> 
> > I believe this repo already has something similar to this as well.
> 
> No, this repo is based on nightfox which I haven't ported these features to yet.

https://github.com/projekt0n/github-nvim-theme/blob/f09a14e9e111f7bfa190dbeab1994b8240dbf2ab/plugin/github_theme.vim#L4-L5

**Comment by nullchilly** [comment]
That doesn't include the lua modules reloading.

**Comment by tmillr** [comment]
> > re-implementing functionality that is already provided by Neovim or LuaJIT (e.g. bit operations, tables to string conversion, hashing, etc.)
> 
> Where is the hashing function in luajit?

I don't believe luajit has one, however, both Vim and Neovim have a sha256 fn `:help sha256()`.

> Also it is an unodered table hashing which (to my knowledge) was never done before in a lua module. It was a result of hours of experimenting on my part.

I see. So maybe it is necessary then (as a dependency for the compile feature). I could be wrong, but I think this sort of order-resistant hashing could be accomplished just by deterministically converting the table to a string (e.g. with `vim.inspect`) and then just sorting the string's lines before hashing it with `sha256()`. There's many ways it could be done.

> > I mean, just look at all the headaches that packer causes due to its custom compilation feature...I see notices/warnings on every other nvim plugin's README warning about it causing problems with their plugin, and now, it seems to be causing similar issues with this plugin too.
> 
> I think it's pretty stable now as there's no issue reported on github or discord for a while. It's a trade-off that @ful1e5 was willing to take between adding more features and performance.

The compiler feature?

I see. Nightfox might actually not have the bug I've been mentioning, but this repo does. The issue is here:

```lua
local git_path = util.join_paths(debug.getinfo(1).source:sub(2, -23), ".git")
```

and has to do with the fact that `github-theme` is 13 characters, while `nightfox` is 9, and there is no `.git` dir in `lua/`. However, I believe there's probably a better way to manipulate paths than doing `string.sub()` and hoping that the filename doesn't change/move.

**Comment by nullchilly** [comment]
> I think this sort of order-resistant hashing could be accomplished just by deterministically converting the table to a string (e.g. with vim.inspect) and then just sorting the string's lines before hashing it with sha256(). There's many ways it could be done

These functions are very, very slow. `vim.inspect` is a debug function which is very, very slow and sha256 is slower than djb2. The current method is **very optimized** with the time of `0.009ms`.

> However, I believe there's probably a better way (instead of just changing the -23 to -27) to manipulate paths than doing string.sub() and hoping that the filename doesn't change/move.

This is just a fast and "reliable" way to get the repo's root. But yeah, PR this change.

> Are these colorschemes intended to be compatible with vim as well? Because I saw some references to vim.api and vim.o and I don't think those are available in standard vim

[Yes.](https://github.com/catppuccin/nvim/blob/main/lua/catppuccin/lib/vim/init.lua)

**Comment by tmillr** [comment]
> > Are these colorschemes intended to be compatible with vim as well? Because I saw some references to vim.api and vim.o and I don't think those are available in standard vim
> 
> [Yes.](https://github.com/catppuccin/nvim/blob/main/lua/catppuccin/lib/vim/init.lua)

Hmmm it seems that polyfill is missing from this repo. I think @ful1e5 intentionally made this one nvim only.

**Comment by tmillr** [comment]
While #270 appears to have fixed the `.git` path bug (and I am no longer seeing a `*/lua/.git` path appearing within my `~/.cache/nvim/github-theme/cache` file), I'm still having stale cache issues where the plugin is not reacting to changes (e.g. I've tried both `touch`ing the `.git` dir after making uncommitted changes, and even completely switching branches). It's still broken. Just now confirmed again that the only way I'm able to get it to work is by doing something like `rm -r ~/.cache/nvim/github-theme && nvim`, and therefore I am keeping this issue open and would like to hear from other users about if they are experiencing a similar issue.

**Comment by ful1e5** [comment]
@tmillr 
> Hmmm it seems that polyfill is missing from this repo. I think @ful1e5 intentionally made this one nvim only.

The reason Vim support is still included in the codebase is because it is built on Nightfox, which also has support for Vim and Neovim. During the refactoring process, I had to release the major version quickly as I had already delayed the release for too long. My intention was to have a strong foundation of theme structure to work with while retaining the previously supported features. Hence, I chose not to remove certain Vim implementations.

There are a few key reasons why I decided against providing full Vim support. First, Vim has its own distinct ecosystem and a range of plugins specifically designed for it. Including full Vim support would require extensive effort to ensure compatibility with these plugins. Additionally, Neovim and Vim are evolving in different directions, which could potentially result in complications in the future.

**Comment by tmillr** [comment]
After doing some debugging recently, I think I stumbled upon something which may be the cause for the stale-cache issues that I'm experiencing (although I'd need to investigate more to be certain).

The final hash is a combination/concatenation of 2 parts, the hash of all settings (incl options, overrides, etc.) and then the literal mtime of the `.git` dir/file. Basically, when `.git` is not detected or doesn't exist (and it is impossible to derive the mtime part), the current impl falls-back to using the literal path to the `.git` dir/file instead as the mtime part. In theory this works nicely for nix setups where there is no `.git` dir and the absolute path to the project/repo dir will contain a unique hash based on the dir's contents. However, one issue with the current impl is that `debug.getinfo().source` can (and often does?) return a relative path which ***does not*** contain the aforementioned nix hash. The path used in this case often looks like `./.git`.

What needs to be done is to change the current impl to convert the path from `debug.getinfo().source` to an absolute path (and, perhaps, resolve symlinks as well) ***before*** using it as part of the final hash. This won't add any noticeable overhead either, since `vim.fn.getftime()` is already doing this internally anyway (i.e. resolving symlinks).

I'm already working on a fix for this here:

https://github.com/projekt0n/github-nvim-theme/pull/290/files#diff-0bbbe30e33a66c696be3c08d6ee0e1c170dabca630ca6ae614f9a21994b8e748R98

as part of a pr which is addressing a separate compiler issue (although perhaps I will extract it out into its own pr instead).

**Comment by tmillr** [comment]
This appears to be fixed at this point. 👍 (see #332 for a couple remaining edge-cases)

The only other thing that I've noticed causing strange, hard-to-track-down bugs has to do with closures (captured values) and/or the use of `Color.BG` (and `C.alpha_blend()`, which uses `Color.BG`), but that is not technically related to the compiler. These global-ish fields are set when a theme's `palette/*.lua` file is required (and, only the first time), which happens for all themes during compilation (but there's only 1 `Color.BG`). At one point, this was causing issues with airline and some of the bg colors used for blending, but I'm unsure if this is still the case. Also, `Color.*` fields don't get set if there's a cache hit and compilation does not occur (which might be a small inconvenience for any downstream users using the `Color` module directly).
